src/number_theory/bernoulli_polynomials.lean

```diff
@@ -0,0 +1,199 @@
+/-
+Copyright (c) 2021 Ashvni Narayanan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Ashvni Narayanan
+-/
+
+import number_theory.bernoulli
+
+/-!
+# Bernoulli polynomials
+
+The Bernoulli polynomials are an important tool obtained from Bernoulli numbers.
+
+## Mathematical overview
+
+The $n^ℐrm{th}$ Bernoulli polynomial is defined as
+$$ B_n(X) = ∑_{k = 0}^n {n χoose k} (-1)^k * B_k * X^{n - k} $$
+where $B_k$ is the $k^ℐrm{th}$ Bernoulli number. The Bernoulli polynomials are generating functions,
+$$ t * e^{tX} / (e^t - 1) = ∑_{n = 0}^{∞} B_n(X) * ¼{t^n}{n!} $$
+
+## Implementation detail
+
+The Bernoulli (negative) numbers are first defined, `bernoulli_neg`, in order to make proving
+theorems using Bernoulli polynomials easier.
+
+## Main theorems
+
+- `sum_bernoulli_poly`: The sum of the $k^\mathrm{th}$ Bernoulli polynomial with binomial
+  coefficients up to n is `(n + 1) * X^n`.
+- `exp_bernoulli_poly`: The Bernoulli polynomials act as generating functions for the exponential.
+-/
+
+noncomputable theory
+open_locale big_operators
+open_locale nat
+
+namespace finset
+
+/-- The two ways of summing over `(i,j)` in the range `a<=i<=j<b` are equal. -/
+lemma sum_Ico_Ico_comm {M : Type*} [add_comm_monoid M]
+  (a b : ℕ) (f : ℕ → ℕ → M) :
+  ∑ i in finset.Ico a b, ∑ j in finset.Ico i b, f i j =
+  ∑ j in finset.Ico a b, ∑ i in finset.Ico a (j+1), f i j :=
+begin
+  rw [finset.sum_sigma', finset.sum_sigma'],
+  refine finset.sum_bij'
+    (λ (x : Σ (i : ℕ), ℕ) _, (⟨x.2, x.1⟩ : Σ (i : ℕ), ℕ)) _ (λ _ _, rfl)
+    (λ (x : Σ (i : ℕ), ℕ) _, (⟨x.2, x.1⟩ : Σ (i : ℕ), ℕ)) _
+    (by rintro ⟨⟩ _; refl) (by rintro ⟨⟩ _; refl);
+  simp only [finset.Ico.mem, sigma.forall, finset.mem_sigma];
+  rintros a b ⟨⟨h₁,h₂⟩, ⟨h₃, h₄⟩⟩; refine ⟨⟨_, _⟩, ⟨_, _⟩⟩; linarith
+end
```

Comment:
This obviously belongs in `finset` somewhere, either near `Ico` or `sum`. If I were you I'd create a new PR with just this lemma to put this in the right place - it saves you waiting for long build times in this PR, and should be merged quickly.

Reply:
Done! 
https://github.com/leanprover-community/mathlib/pull/6361 -- for the double sum